packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,83 @@
+import { isValidElement } from 'react';
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL, SLOT_COMPONENT_TYPEOF_SYMBOL } from './constants';
+import * as React from 'react';
+import { SlotPropsRecord } from './types';
+import { ExtractSlotProps } from './types';
+import { PropsWithoutRef } from './types';
+import { getNativeElementProps } from '../utils/getNativeElementProps';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  elementType:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+};
+
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  options: { shorthand: Props | SlotShorthandValue | undefined; required: true } & SlotOptions<Props>,
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  options: { shorthand: Props | SlotShorthandValue | undefined | null; required?: boolean } & SlotOptions<Props>,
+): SlotComponent<Props> | undefined;
+export function slot<Props extends UnknownSlotProps = UnknownSlotProps>(
+  options: { shorthand: Props | SlotShorthandValue | undefined | null; required?: boolean } & SlotOptions<Props>,
+): SlotComponent<Props> | undefined {
```

Comment:
It looks like this `slot` function is saving the render function in a different way than `resolveShorthand` (on a metadata object vs. on `SLOT_RENDER_FUNCTION_SYMBOL`). 

That's fine in general, but it would be nice to keep interoperability between calls to `resolveShorthand` and `slot`. The main reason is that `resolveShorthand` is the only way for a component to modify slot props of a different component. E.g. if a Foo component has a slot of type Bar, and wants to modify the props sent to its `bar` slot, it would need to call `resolveShorthand` on the `bar` slot first. Here's an example where a component needs to modify the slot of another component: https://github.com/microsoft/fluentui/pull/27834

Here's a contrived stress test case that should work. Multiple repeated calls to `resolveShorthand` followed by a call to `slot`:

```tsx
const component = <TestComponent mySlot={{ children: () => { /* render function */ }} />;

// Inside TestComponent:

// The first call moves the render function to `SLOT_RENDER_FUNCTION_SYMBOL`:
const mySlotResolved1 = resolveShorthand(props.mySlot);
// The second call should maintain the render function.
const mySlotResolved2 = resolveShorthand(mySlotResolved1);
// The call to slot should also maintain the render function.
const mySlot = slot({ shorthand: mySlotResolved2 });

// When rendering the component with mySlot, it should use the render function.
```


I think in order for that to work, `slot` would need to check for the existence of `SLOT_RENDER_FUNCTION_SYMBOL` and make sure it ends up in the right place on the metadata.